packages/filter-workspace-packages/src/index.ts

```diff
@@ -66,13 +70,54 @@ export async function filterPkgsBySelectorObjects<T> (
     selectedProjectsGraph: PackageGraph<T>
     unmatchedFilters: string[]
   }> {
-  const { graph } = createPkgGraph<T>(pkgs, { linkWorkspacePackages: opts.linkWorkspacePackages })
-  if (packageSelectors?.length) {
-    return filterGraph(graph, packageSelectors, {
-      workspaceDir: opts.workspaceDir,
-      testPattern: opts.testPattern,
+  packageSelectors = packageSelectors.filter((selector) => !selector.followProdDepsOnly)
+  const prodPackageSelectors = packageSelectors.filter((selector) => selector.followProdDepsOnly)
+
+  if ((packageSelectors.length > 0) || (prodPackageSelectors.length > 0)) {
+    let filteredGraph = Promise.resolve<{
```

Comment:
I am not sure this optimization with promises is important. Will people frequently mix `--filter-prod` with `--filter`? anyway, refactor it to use a type:

```
let filteredGraph: FilteredGraph
```